### PR TITLE
sd-boot: don't read TSC in VM environments, and other TSC fixes

### PR DESCRIPTION
**Pull request by poettering**
Replaces: #22388

### CODE REVIEW

File: src/boot/efi/ticks.c [hunk 1]
```diff
@@ -2,18 +2,47 @@
...
+#include <cpuid.h>
+#endif
 
 #include "ticks.h"
 
+#if defined(__i386__) || defined(__x86_64__)
+static BOOLEAN in_hypervisor(void) {
+        uint32_t eax, ebx, ecx, edx;
+
+        /* The TSC might or might not be virtualized in VMs (an thus might not be accurate or start at zero
```
Comment (yuwata) [inline src/boot/efi/ticks.c]: an thus -> and thus?

**Review by yuwata — APPROVED** [review]
LGTM. One possible typo.

**Comment by poettering** [comment]
Force pushed new version with the typo fixed.

File: src/boot/efi/util.c [hunk 2]
```diff
@@ -3,69 +3,9 @@
 #include <efi.h>
 #include <efilib.h>
 
+#include "ticks.h"
```
Comment (medhefgo) [inline src/boot/efi/util.c:6]: It would be better header hygiene to include it where it's actually needed.

File: src/boot/efi/ticks.c [hunk 3]
```diff
@@ -45,6 +45,10 @@ UINT64 ticks_freq(void) {
         BS->Stall(1000);
         ticks_end = ticks_read();
 
+        if (ticks_end < ticks_start) /* Check for an overflow (which is not that unlikely, given on some
+                                      * archs the value is 32bit) */
```
Comment (medhefgo) [inline src/boot/efi/ticks.c:81]: That's some weird formatting…

**Review by medhefgo — APPROVED** [review]
If it works…

It's a shame the EFI timestamp protocol is utter garbage to use. I looked at this some months ago and it overflows within seconds on real hardware. Other similar protocols are just as useless :cry:.

File: src/boot/efi/util.c (hunk 2 above)
Comment (poettering) [reply src/boot/efi/util.c:6]: efivar_set_time_usec() further down in util.c uses time_usec(), that's why we include it here.

File: src/boot/efi/ticks.c (hunk 3 above)
Comment (poettering) [reply src/boot/efi/ticks.c:81]: It's how emacs line breaks this, and Emacs is God. Of course...

**Pull request merged**